File: renard2024ros/part-ii/ros-workspace/src/cpppkg/src/f.cpp
#include "my_interfaces/srv/reset_counter.hpp"
#include "rclcpp/rclcpp.hpp"

using ResetCounter = my_interfaces::srv::ResetCounter;

class Client : public rclcpp::Node {
  // The size of the ROS-based queue.
  //
  // This is a static variable used to set the queue size of ROS-related
  // publishers, accordingly.
  static const int qsize = 10;

public:
  Client() : Node("f") {
    // Create the client(s).
    //
    // This will create the set of client(s) that this node is responsible for
    // handling, accordingly.
    this->clt = this->create_client<ResetCounter>("reset");
  }

  // Make a request.
  //
  // This sends a request to the service to set the counter value to a specified
  // amount, accordingly.
  auto request(int value) -> void {
    while (!this->clt->wait_for_service(std::chrono::seconds(1))) {
      RCLCPP_WARN(this->get_logger(), "waiting for server...");
    }

    auto request = std::make_shared<ResetCounter::Request>();
    request->reset = value;

    // Send the request.
    //
    // This will send an asynchronous call to the server to have its request
    // fulfilled.
    this->clt->async_send_request(
        request, std::bind(&Client::callback, this, std::placeholders::_1));
  }

  auto callback(rclcpp::Client<ResetCounter>::SharedFuture future) -> void {
    // Retrieve the result.
    //
    // The result (i.e., the data) of the response is retrieved by calling the
    // `get` method on the value, accordingly.
    auto response = future.get();

    RCLCPP_INFO(this->get_logger(), "response: `success=%d`",
                response->success);
    RCLCPP_INFO(this->get_logger(), "response: `message=%s`",
                response->message.c_str());
  }

private:
  rclcpp::Client<ResetCounter>::SharedPtr clt;
};

int main(int argc, char **argv) {
  rclcpp::init(argc, argv);
  auto node = std::make_shared<Client>();

  // Submit request.
  //
  // This submits the request a single time after the node has been successfully
  // created.
  node->request(10);

  // Spin-up the ROS-based node.
  //
  // This will run the ROS-styled node infinitely until the signal to stop the
  // program is received, accordingly.
  rclcpp::spin(node);

  // Shut the node down, gracefully.
  //
  // This will close and exit the node execution without disrupting the ROS
  // communication network, assumingly.
  rclcpp::shutdown();

  // The final return.
  //
  // This is required for the main function of a program within the C++
  // programming language.
  return 0;
}

File: renard2024ros/part-ii/src/pypkg/pypkg/f.py
import rclpy

from my_interfaces.srv import ResetCounter
from rclpy.node import Node

class Client(Node):
    """A node to represent a client for resetting the counter.
    """

    def __init__(self):
        """Initialize the node.
        """

        super().__init__("f")
        self.logger = self.get_logger()

        # Create client(s).
        #
        # This is where all the client(s) of this node are created and managed,
        # accordingly.
        self.clt = self.create_client(ResetCounter, "reset")

    def request(self, value: int):
        """Make a request to the service.
        """

        while not self.clt.wait_for_service(1.0):
            self.logger.info("waiting for service...")

        # Create the request.
        #
        # At this point, we can create the interface needed for the service to
        # be called, accordingly.
        request = ResetCounter.Request()
        request.reset = value

        # Make request.
        #
        # This will make an asynchronous call to the service and wait for a
        # response.
        future = self.clt.call_async(request)

        # Add a callback to the request.
        #
        # This callback is responsible for being ran when the response gets
        # fulfilled, accordingly.
        future.add_done_callback(self.callback)

    def callback(self, future):
        """A callback for responses from the service.
        """

        # Retrieve the result.
        #
        # The result (i.e., the data) of the response is retrieved by calling the
        # `result` on the value.
        response = future.result()

        self.logger.info(f"reponse: `success={response.success}`")
        self.logger.info(f"reponse: `message={response.message}`")
        

def main(args=None):
    """The main function.
    """

    rclpy.init(args=args)
    node = Client()

    node.request(10)

    # Spin-up the node.
    #
    # This will start the node up, accordingly. This function call is also thread
    # blocking until a signal stop is sent to it, accordingly.
    rclpy.spin(node)

    # Shut this node down, gracefully.
    #
    # This will shut the node down from the ROS communication statck without
    # causing any major hiccups, assumingly.
    rclpy.shutdown()
